crates/shared/src/price_estimation/quasimodo.rs

```diff
@@ -0,0 +1,191 @@
+use ethcontract::{H160, U256};
+use gas_estimation::GasPriceEstimating;
+use model::order::OrderKind;
+use model::TokenPair;
+use num::{BigInt, BigRational};
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+pub struct QuasimodoPriceEstimator {
+    pub api: Arc<HttpSolverApi>,
+    pub pools: Arc<PoolCache>,
+    pub bad_token_detector: Arc<dyn BadTokenDetecting>,
+    pub token_info: Arc<dyn TokenInfoFetching>,
+    pub gas_info: Arc<dyn GasPriceEstimating>,
+    pub native_token: H160,
+    pub base_tokens: Arc<BaseTokens>,
+}
+
+impl QuasimodoPriceEstimator {
+    async fn estimate(&self, query: &Query) -> Result<Estimate, PriceEstimationError> {
+        if query.buy_token == query.sell_token {
+            return Ok(Estimate {
+                out_amount: query.in_amount,
+                gas: 0.into(),
+            });
+        }
+
+        ensure_token_supported(query.buy_token, self.bad_token_detector.as_ref()).await?;
+        ensure_token_supported(query.sell_token, self.bad_token_detector.as_ref()).await?;
+
+        let gas_cost = U256::from_f64_lossy(self.gas_info.estimate().await?.legacy);
+
+        let mut tokens = self.base_tokens.tokens().clone();
+        tokens.insert(query.sell_token);
+        tokens.insert(query.buy_token);
+        tokens.insert(self.native_token);
+        let tokens: Vec<_> = tokens.drain().collect();
+
+        let token_infos = self.token_info.get_token_infos(&tokens).await;
+
+        let tokens = self
+            .base_tokens
+            .tokens()
+            .iter()
+            .map(|token| {
+                (
+                    *token,
+                    TokenInfoModel {
+                        decimals: token_infos[token].decimals,
+                        normalize_priority: Some(if self.native_token == query.buy_token {
+                            1
+                        } else {
+                            0
+                        }),
+                        ..Default::default()
+                    },
+                )
+            })
+            .collect();
+
+        let (sell_amount, buy_amount) = match query.kind {
+            OrderKind::Buy => (U256::max_value(), query.in_amount),
+            OrderKind::Sell => (query.in_amount, U256::one()),
+        };
+
+        let orders = BTreeMap::from([(
+            0,
+            OrderModel {
+                sell_token: query.sell_token,
+                buy_token: query.buy_token,
+                sell_amount,
+                buy_amount,
+                allow_partial_fill: false,
+                is_sell_order: query.kind == OrderKind::Sell,
+                fee: FeeModel {
+                    amount: *GAS_PER_ORDER * gas_cost,
+                    token: self.native_token,
+                },
+                cost: CostModel {
+                    amount: *GAS_PER_ORDER * gas_cost,
+                    token: self.native_token,
+                },
```

Comment:
Neat idea to make it so the order cost doesn't matter.

Just wanted to confirm with @marcovc that its OK to have fee in a token other than the sell token.

Reply:
Yes, it is fine.